securesystemslib/signer/_hsm_signer.py

```diff
@@ -0,0 +1,301 @@
+"""Hardware Security Module (HSM) Signer
+
+Uses PKCS#11/Cryptoki API to create signatures with HSMs (e.g. YubiKey) and to export
+the related public keys.
+
+"""
+import binascii
+from contextlib import contextmanager
+from typing import Optional, Tuple
+from urllib import parse
+
+from securesystemslib import KEY_TYPE_ECDSA
+from securesystemslib.exceptions import UnsupportedLibraryError
+from securesystemslib.signer._key import Key, SSlibKey
+from securesystemslib.signer._signature import Signature
+from securesystemslib.signer._signer import SecretsHandler, Signer
+
+# pylint: disable=wrong-import-position
+CRYPTO_IMPORT_ERROR = None
+try:
+    from cryptography.hazmat.primitives import serialization
+    from cryptography.hazmat.primitives.asymmetric.ec import (
+        SECP256R1,
+        SECP384R1,
+        EllipticCurvePublicKey,
+        ObjectIdentifier,
+        get_curve_for_oid,
+    )
+    from cryptography.hazmat.primitives.asymmetric.utils import (
+        encode_dss_signature,
+    )
+
+    # TODO: Don't hardcode schemes
+    _SCHEME_FOR_CURVE = {
+        SECP256R1: "ecdsa-sha2-nistp256",
+        SECP384R1: "ecdsa-sha2-nistp384",
+    }
+    _CURVE_NAMES = [curve.name for curve in _SCHEME_FOR_CURVE]
+
+except ImportError:
+    CRYPTO_IMPORT_ERROR = "'cryptography' required"
+
+PYKCS11_IMPORT_ERROR = None
+try:
+    from PyKCS11 import PyKCS11
+
+    # TODO: Don't hardcode schemes
+    _MECHANISM_FOR_SCHEME = {
+        "ecdsa-sha2-nistp256": PyKCS11.Mechanism(PyKCS11.CKM_ECDSA_SHA256),
+        "ecdsa-sha2-nistp384": PyKCS11.Mechanism(PyKCS11.CKM_ECDSA_SHA384),
+    }
+
+except ImportError:
+    PYKCS11_IMPORT_ERROR = "'PyKCS11' required"
+
+ASN1_IMPORT_ERROR = None
+try:
+    from asn1crypto.keys import (  # pylint: disable=import-error
+        ECDomainParameters,
+        ECPoint,
+    )
+except ImportError:
+    ASN1_IMPORT_ERROR = "'asn1crypto' required"
+# pylint: enable=wrong-import-position
+
+
+_PYKCS11LIB = None
+
+
+def PYKCS11LIB():
```

Comment:
Question for reviewers: Any better ideas than this? See 6b825a0ee4aea04107f9354b91c566ac7ed52294 for context.